org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java

```diff
@@ -637,30 +643,106 @@ private void collectionRegionProposals(ICompletionRequest request, ICompletionRe
 			String lineUntilPos = lineText.substring(0, pos.getCharacter());
 			Matcher match = regionCompletionRegExpr.matcher(lineUntilPos);
 			if (match.find()) {
+				InsertTextFormat insertFormat = request.getInsertTextFormat();
 				Range range = new Range(new Position(pos.getLine(), pos.getCharacter() + match.regionStart()), pos);
 
+				String text = request.isCompletionSnippetsSupported() ? "<!-- #region $1-->" : "<!-- #region -->";
 				CompletionItem beginProposal = new CompletionItem("#region");
-				beginProposal.setTextEdit(new TextEdit(range, "<!-- #region $1-->"));
-				beginProposal.setDocumentation("Folding Region Start");
+				beginProposal.setTextEdit(new TextEdit(range, text));
+				beginProposal.setDocumentation("Insert Folding Region Start");
 				beginProposal.setFilterText(match.group());
 				beginProposal.setSortText("za");
 				beginProposal.setKind(CompletionItemKind.Snippet);
-				beginProposal.setInsertTextFormat(InsertTextFormat.Snippet);
+				beginProposal.setInsertTextFormat(insertFormat);
 				response.addCompletionAttribute(beginProposal);
 
 				CompletionItem endProposal = new CompletionItem("#endregion");
-				endProposal.setKind(CompletionItemKind.Snippet);
 				endProposal.setTextEdit(new TextEdit(range, "<!-- #endregion-->"));
-				endProposal.setDocumentation("Folding Region End");
+				endProposal.setDocumentation("Insert Folding Region End");
 				endProposal.setFilterText(match.group());
 				endProposal.setSortText("zb");
+				endProposal.setKind(CompletionItemKind.Snippet);
+				endProposal.setInsertTextFormat(InsertTextFormat.PlainText);
 				response.addCompletionAttribute(endProposal);
 			}
 		} catch (BadLocationException e) {
 			LOGGER.log(Level.SEVERE, "While performing collectRegionCompletion", e);
 		}
 	}
 
+	private void collectCDATACompletion(ICompletionRequest request, ICompletionResponse response) {
+		try {
+			boolean isSnippetsSupported = request.isCompletionSnippetsSupported();
+			InsertTextFormat insertFormat = request.getInsertTextFormat();
+
+			DOMDocument document = request.getXMLDocument();
+
+			String filter = "<!CDATA";
+			int startOffset = StringUtils.findExprBeforeAt(document.getText(), filter, request.getOffset());
+			if (startOffset == -1) {
+				startOffset = StringUtils.findExprBeforeAt(document.getText(), "CDATA", request.getOffset());
+				if (startOffset != - -1) {
+					filter = "CDATA";
+				}
+			}
+			if (startOffset == -1) {
+				startOffset = request.getOffset();
+			}
+			int endOffset = request.getOffset();
+			Range editRange = getReplaceRange(startOffset + 1, endOffset, request);
+
+			// Insert CDATA Section
+			CompletionItem cdataProposal = new CompletionItem("cdata");
+			cdataProposal.setKind(CompletionItemKind.Snippet);

```

Comment:
isSnippetsSupported?

Reply:
It's just for icon. Snippet support is for insert text format